pkg/shared/utils/sync/distributed_lock_mgr_test.go

```diff
@@ -0,0 +1,203 @@
+package sync
+
+import (
+	"fmt"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/shared/utils/arrays"
+	"github.com/onsi/gomega"
+	"gorm.io/gorm"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+var _ DistributedLock = &__distributedLockMock{}
+
+// this map is used to simulate database locks
+var mutexMap map[string]*sync.Mutex
+
+// this lock is used to sync access to the mutexMap
+var mu sync.Mutex
+
+func getMutex(lockID string) *sync.Mutex {
+	mu.Lock()
+	defer mu.Unlock()
+	if mutexMap == nil {
+		mutexMap = map[string]*sync.Mutex{}
+	}
+
+	testLock, ok := mutexMap[lockID]
+	if !ok {
+		testLock = &sync.Mutex{}
+		mutexMap[lockID] = testLock
+	}
+	return testLock
+}
+
+type __distributedLockMock struct {
+	lockID string
+	locked bool
+}
+
+func (d *__distributedLockMock) isLocked() bool {
+	return d.locked
+}
+
+func (d *__distributedLockMock) Lock() error {
+	mu := getMutex(d.lockID)
+	mu.Lock()
+	d.locked = true
+	return nil
+}
+
+func (d *__distributedLockMock) Unlock() error {
+	getMutex(d.lockID).Unlock()
+	d.locked = false
+	return nil
+}
+
+func createDistributedLock(db *gorm.DB, lockID string) DistributedLock {
+	return &__distributedLockMock{
+		lockID: lockID,
+	}
+}
+
+type testRoutineParams struct {
+	wg                     *sync.WaitGroup
+	distributedLockManager DistributedLockMgr
+	delay                  time.Duration
+	lockID                 string
+	current                *int32
+	dest                   *int32
+	g                      *gomega.WithT
+}
+
+func getNext(p *testRoutineParams) {
+	p.g.Expect(p.distributedLockManager.Lock(p.lockID)).To(gomega.Succeed())
+
+	*p.dest = *p.current + 1
+	time.Sleep(p.delay)
```

Comment:
Do you think there might be a way to perform the tests without depending on sleeping?
The sleep times end up adding between several tests so I would prefer avoiding them if possible. We can create a follow-up jira issue to investigate that if at the moment we are not sure whether it is possible or not

Reply:
We could consider alternative testing methods. Currently, I have used the sleep function to simulate load and ensure that the routines do not finish too quickly before the next one begins.

Reply:
:+1: let's investigate that if possible